File: src/main/java/com/github/timo_reymann/ansible_vault_integration/execution/AnsibleVaultTask.java
package com.github.timo_reymann.ansible_vault_integration.execution;

import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Background task abstraction for executing ansible vault operations
 */
public class AnsibleVaultTask extends Task.Backgroundable {
    private static final Logger log = Logger.getInstance(AnsibleVaultTask.class);
    public static String NOTIFICATION_GROUP_ID = AnsibleVaultTask.class.getCanonicalName();
    public static String NOTIFICATION_ERROR_ID = "Error";
    public static String NOTIFICATION_SUCCESS_ID = "Success";

    private final AnsibleVaultRunnable task;

    public AnsibleVaultTask(@Nullable Project project, @Nls(capitalization = Nls.Capitalization.Title) @NotNull String title, AnsibleVaultRunnable task) {
        super(project, title);
        this.task = task;
    }

    @Override
    public void run(@NotNull ProgressIndicator indicator) {
        indicator.setIndeterminate(true);
        indicator.setText(this.myTitle);
        Notification notification = null;

        try {
            this.task.run();
            notification = new Notification(
                    NOTIFICATION_GROUP_ID,
                    NOTIFICATION_SUCCESS_ID,
                    task.getSuccessMessage(),
                    NotificationType.INFORMATION
            );
        } catch (Exception e) {
            log.error(e);
            notification = new Notification(
                    NOTIFICATION_GROUP_ID,
                    NOTIFICATION_ERROR_ID,
                    e.getMessage(),
                    NotificationType.ERROR
            );
            indicator.cancel();
        }

        Notifications.Bus.notify(notification);
    }
}

File: src/main/kotlin/de/timo_reymann/ansible_vault_integration/action/execution/AnsibleVaultTask.kt
package de.timo_reymann.ansible_vault_integration.action.execution

import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.notification.Notifications
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.Task.Backgroundable
import com.intellij.openapi.project.Project
import de.timo_reymann.ansible_vault_integration.action.runnable.AnsibleVaultRunnable
import org.jetbrains.annotations.Nls

/**
 * Background task abstraction for executing ansible vault operations
 */
class AnsibleVaultTask(
    project: Project?,
    title: @Nls(capitalization = Nls.Capitalization.Sentence) String,
    private val task: AnsibleVaultRunnable
) : Backgroundable(project, title) {
    override fun run(indicator: ProgressIndicator) {
        indicator.isIndeterminate = true
        indicator.text = this.title
        var notification: Notification
        try {
            task.run()
            notification = Notification(
                NOTIFICATION_GROUP_ID,
                NOTIFICATION_SUCCESS_ID,
                task.successMessage,
                NotificationType.INFORMATION
            )
        } catch (e: Exception) {
            log.warn(e)
            notification = Notification(
                NOTIFICATION_GROUP_ID,
                NOTIFICATION_ERROR_ID,
                e.message!!,
                NotificationType.ERROR
            )
            indicator.cancel()
        }
        Notifications.Bus.notify(notification)
    }

    companion object {
        private val log = Logger.getInstance(AnsibleVaultTask::class.java)

        val NOTIFICATION_GROUP_ID = AnsibleVaultTask::class.java.canonicalName
        const val NOTIFICATION_ERROR_ID = "Error"
        const val NOTIFICATION_SUCCESS_ID = "Success"
    }
}
